core/src/main/java/org/infinispan/notifications/cachelistener/CacheNotifierImpl.java

```diff
@@ -301,23 +294,36 @@ public void setTypeConverter(TypeConverter typeConverter) {
       return allowedListeners;
    }
 
-   private K convertKey(K key) {
-      if (typeConverter != null) {
-         key = (K) typeConverter.unboxKey(key);
-      }
-      if (cacheConverter != null) {
-         key = cacheConverter.unboxKey(key);
+   private Encoder getEncoder(Encoder invocationEncoder) {
+      return invocationEncoder != null ? invocationEncoder : encoder;
+   }
+
+   private Wrapper getWrapper(Wrapper invocationWrapper) {
+      return invocationWrapper != null ? invocationWrapper : wrapper;
+   }
+
+   private K convertKey(Encoder encoder, Wrapper wrapper, K key) {
+      Encoder enc = getEncoder(encoder);
+      Wrapper wrap = getWrapper(wrapper);
+
+      if (!enc.isStorageFormatFilterable()) {
+         key = (K) getEncoder(encoder).keyFromStorage(getWrapper(wrapper).unwrapKey(key));
```

Comment:
Seems we can reduce this if/else block to
`key = (K) wrap.unwrapKey(key);`
`if (!enc.isStorageFormatFilterable()) {
   key = (K) getEncoder(encoder).keyFromStorage(key);
}
`